### DISCUSSION TITLE
How do I navigate from find_timing_paths to the ODB objects?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#6072] [General]
I would like to create a 3d graph in python for megaboom to see the skew as a function of x and y on the design as I was inspired by slide 20 in https://www.eng.biu.ac.il/temanad/files/2017/02/Lecture-8-CTS.pdf

(Of course this is just a heatmap, the 3d bit is just bling, but helps create a sense of what is going on)

Here is a first attempt, using slack instead of skew, I don't know how to get from a timing path to skew, although the OpenROAD GUI shows skew per path. Though from the slack and x,y I do get a sense of what is going on.

![image](https://github.com/user-attachments/assets/54b6ab0f-49b5-4993-83ec-007070bd21ca)

However, I haven't found a reasonable way to map a skew to an x,y location. The transition from find_timing_paths to the ODB objects is particularly awkward/hacky below...

Any pointers on how to perfom going from find_timing_slack to skew & x,y position in tcl?

```
source $::env(SCRIPTS_DIR)/load.tcl
load_design 4_cts.odb 4_cts.sdc

set paths [find_timing_paths -path_group reg2reg -sort_by_slack -group_count 1000]

set db [::ord::get_db]
set block [[$db getChip] getBlock]

# open file
set fp [open $::env(OUTPUT) w]
foreach path $paths {
  set endpoint [get_property $path endpoint]
  set instance [$endpoint instance]
  set name [get_property $endpoint full_name]
  set slack [get_property $path slack]

  # If DFF is in the $name and it ends with /D
  if {[regexp {DFF} $name] && [regexp {/D$} $name]} {
    # remove the /D suffix
    set name [string range $name 0 end-2]
    # replace [ and ] with \[ and \]
    set name [string map { {[} {\[} {]} {\]} } $name]
    set b [$block findInst $name]
    set bbox [$b getBBox]
    puts $fp "$name $slack [$bbox xMin] [$bbox yMin]"
  }
}
close $fp
```

**Comment by oharboe** [comment]
Seen from this angle, it is effectively a heatmap. This is with 10000 WNS paths. This is corroborating that [global placement merits further study](https://github.com/The-OpenROAD-Project/OpenROAD/issues/6066).

Tinkering a bit more and checking that I'm getting roughly the right coordinate, I can see that the WNS is in the same location in the plot and in OpenROAD GUI

![image](https://github.com/user-attachments/assets/8a2b43ed-6014-42cd-8b5f-fdc66fdcdf57)

![image](https://github.com/user-attachments/assets/54326de2-d358-4410-ae7f-a1cb218e269c)


Bumped to 1000000 paths:

![image](https://github.com/user-attachments/assets/f31a2a40-d967-4e22-b12f-18a159581fe8)

**Comment by maliberty** [comment]
I think you want the sta_to_db_* functions from dbSta.i

**Reply by oharboe** [reply]
Will try, thanks!

Is there a way to get the skew from a path in tcl, same as the GUI?

**Reply by maliberty** [reply]
I don't know of one but you could ask in the sta repo

**Reply by oharboe** [reply]
> I think you want the sta_to_db_* functions from dbSta.i

```
$ make RUN_SCRIPT=slack-positions.tcl run
[deleted]
Expected instance NULL
make: *** [Makefile:1076: run] Error 1
```

```
source $::env(SCRIPTS_DIR)/load.tcl
load_design 4_cts.odb 4_cts.sdc

set paths [find_timing_paths -path_group reg2reg -sort_by_slack -group_count 1000000]

set db [::ord::get_db]
set block [[$db getChip] getBlock]

set path [lindex $paths 0]
set endpoint [get_property $path endpoint]
set instance [sta::sta_to_db_inst [$endpoint instance]]
puts "Expected instance $instance"
set name [get_property $endpoint full_name]
set slack [get_property $path slack]
set bbox [$instance getBBox]
puts "$name $slack [$bbox xMin] [$bbox yMin]"
```

**Comment by maliberty** [comment]
Those plots are labeled DELAY and Skew which shows some inconsistency.  I think they are measuring delay from the driver to a sequential not really a skew (I am guessing).  If you see two points at different heights there is no way to know whether a path exists.

**Reply by oharboe** [reply]
I meant in the Timing Report:

![image](https://github.com/user-attachments/assets/f9bc36a4-8e29-40a3-892f-b97def247879)

**Comment by oharboe** [comment]
Seems to work now, less slack:

```
source $::env(SCRIPTS_DIR)/load.tcl
load_design 4_cts.odb 4_cts.sdc

set paths [find_timing_paths -path_group reg2reg -sort_by_slack -group_count 1000000]

set db [::ord::get_db]
set block [[$db getChip] getBlock]

foreach path $paths {
  set endpoint [get_property $path endpoint]
  set instance [sta::sta_to_db_inst [$endpoint instance]]
  set port [sta::sta_to_db_port [$endpoint port]]
  set name [get_property $endpoint full_name]
  set obj {NULL}
  if {$port!={NULL}} {
    set obj $port
  } elseif {$instance!={NULL}} {
    set obj $instance
  } else {
    puts "Nothing found for $name"
    continue
  }
  set slack [get_property $path slack]
  set bbox [$obj getBBox]
  puts "$name $slack [ord::dbu_to_microns [$bbox xMin]] [ord::dbu_to_microns [$bbox yMin]]"
}
```

I get:

```
$ make RUN_SCRIPT=slack-positions.tcl run
/home/oyvind/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad -exit -no_init -threads 16  -no_splash  slack-positions.tcl
[WARNING STA-0527] unknown path group 'reg2reg'.
resp_msg[15] -0.029667 21.305 36.24
dpath.a_reg.out[7]$_DFFE_PP_/D -0.028715 31.35 18.2
[deleted]
req_rdy 0.194209 0.0 12.88
```
